spring-integration-mqtt/src/main/java/org/springframework/integration/mqtt/inbound/Mqttv5PahoMessageDrivenChannelAdapter.java

```diff
@@ -187,33 +190,40 @@ protected void doStart() {
 	@Override
 	protected void doStop() {
 		this.topicLock.lock();
-		String[] topics = getTopic();
+		initializeMqttAsyncClientIfRequired();
+		String[] topics = getTopicNames();
 		try {
 			this.mqttClient.unsubscribe(topics).waitForCompletion(getCompletionTimeout());
 			this.mqttClient.disconnect().waitForCompletion(getCompletionTimeout());
-		}
-		catch (MqttException ex) {
+		} catch (MqttException ex) {
 			logger.error(ex, () -> "Error unsubscribing from " + Arrays.toString(topics));
-		}
-		finally {
+		} finally {
 			this.topicLock.unlock();
 		}
 	}
 
+	private void initializeMqttAsyncClientIfRequired() {
+		if (mqttClient == null) {
+			initializeMqttAsyncClient();
+		}
+	}
+
 	@Override
 	public void destroy() {
 		super.destroy();
 		try {
-			this.mqttClient.close(true);
-		}
-		catch (MqttException ex) {
+			if (mqttClient != null) {
+				this.mqttClient.close(true);
+			}
+		} catch (MqttException ex) {
 			logger.error(ex, "Failed to close 'MqttAsyncClient'");
 		}
 	}
 
 	@Override
 	public void addTopic(String topic, int qos) {
 		this.topicLock.lock();
+		initializeMqttAsyncClientIfRequired();
```

Comment:
That's not OK, too. 
The `addTopic()` could be called from the Spring context initialization phase. Therefore it is too aggressive to make a low-level resource connection.

I believe the logic here must be similar to the one in the `MqttPahoMessageDrivenChannelAdapter`:
```
if (this.client != null && this.client.isConnected()) {
		this.client.subscribe(topic, qos);
}
```